vendors/swedbank-pay-core-library/src/SwedbankPay/Core/Library/OrderAction.php

```diff
@@ -0,0 +1,573 @@
+                throw new Exception($message);
+            default:
+                throw new Exception('Capture is failed.');
+        }
+
+        return $result;
+    }
+
+    /**
+     * Cancel.
+     *
+     * @param mixed $orderId
+     * @param mixed $amount
+     * @param mixed $vatAmount
+     *
+     * @return Response
+     * @throws Exception
+     */
+    public function cancel($orderId, $amount = null, $vatAmount = 0)
+    {
+        /** @var Order $order */
+        $order = $this->getOrder($orderId);
+
+        if (!$amount) {
+            $amount = $order->getAmount();
+            $vatAmount = $order->getVatAmount();
+        }
+
+        if (!$this->canCancel($orderId, $amount)) {
+            throw new Exception('Cancellation is not available.');
+        }
+
+        $paymentId = $order->getPaymentId();
+        if (empty($paymentId)) {
+            throw new Exception('Unable to get payment ID');
+        }
+
+        // Checkout method can use the Invoice method
+        $info = $this->fetchPaymentInfo($paymentId);
+        if ($info['payment']['instrument'] === 'Invoice') {
+            // @todo Should we use different credentials?
+            return $this->cancelInvoice($orderId, $amount, $vatAmount);
+        }
+
+        $href = $info->getOperationByRel('create-cancellation');
+        if (empty($href)) {
+            throw new Exception('Cancellation is unavailable');
+        }
+
+        $params = [
+            'transaction' => [
+                'description' => sprintf('Cancellation for Order #%s', $order->getOrderId()),
+                'payeeReference' => $this->generatePayeeReference($orderId)
+            ],
+        ];
+        $result = $this->request('POST', $href, $params);
+
+        // Save transaction
+        $transaction = $result['cancellation']['transaction'];
+        $this->saveTransaction($orderId, $transaction);
+
+        switch ($transaction['state']) {
+            case 'Completed':
+                $this->updateOrderStatus(OrderInterface::STATUS_CANCELLED, 'Transaction is cancelled.');
+                break;
+            case 'Initialized':
+            case 'AwaitingActivity':
+                $this->updateOrderStatus(OrderInterface::STATUS_CANCELLED,
+                    sprintf('Transaction cancellation status: %s.', $transaction['state']));
+                break;
+            case 'Failed':
+                $message = isset($transaction['failedReason']) ? $transaction['failedReason'] : 'Cancellation is failed.';
+                throw new Exception($message);
+            default:
+                throw new Exception('Capture is failed.');
+        }
+
+        return $result;
+    }
+
+    /**
+     * Refund.
+     *
+     * @param mixed $orderId
+     * @param mixed $amount
+     * @param mixed $vatAmount
+     *
+     * @return Response
+     * @throws Exception
+     */
+    public function refund($orderId, $amount = null, $vatAmount = 0)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method refund() has a Cyclomatic Complexity of 11. The configured cyclomatic complexity threshold is 10.](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-woocommerce-payments/pullRequest?prid=5340032)